plugin/src/main/java/org/opensearch/ml/action/controller/DeleteModelControllerTransportAction.java

```diff
@@ -0,0 +1,249 @@
+                                        modelId
+                                    );
+                                deleteModelController(modelId, actionListener);
+                            } else {
+                                String[] nodeIds = getUndeployModelControllerFailedNodesList(nodesResponse);
+                                log
+                                    .error(
+                                        "Failed to undeploy model controller with model ID {} on following nodes {}, deletion is aborted. Please retry or undeploy the model manually and then perform the deletion.",
+                                        modelId,
+                                        Arrays.toString(nodeIds)
+                                    );
+                                actionListener
+                                    .onFailure(
+                                        new RuntimeException(
+                                            "Failed to undeploy model controller with model ID "
+                                                + modelId
+                                                + " on following nodes "
+                                                + Arrays.toString(nodeIds)
+                                                + ", deletion is aborted. Please retry or undeploy the model manually and then perform the deletion."
+                                        )
+                                    );
+                            }
+                        }, e -> {
+                            log
+                                .error(
+                                    "Failed to undeploy model controller from cache and delete the model controller for model {}",
+                                    modelId,
+                                    e
+                                );
+                            actionListener.onFailure(e);
+                        }), context::restore)
+                    );
+            } else {
+                deleteModelController(modelId, actionListener);
+            }
+        } catch (Exception e) {
+            log.error("Failed to delete model controller", e);
+            actionListener.onFailure(e);
+        }
+    }
+
+    private void deleteModelController(String modelId, ActionListener<DeleteResponse> actionListener) {
+        DeleteRequest deleteRequest = new DeleteRequest(ML_MODEL_CONTROLLER_INDEX, modelId);
+        client.delete(deleteRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(DeleteResponse deleteResponse) {
+                log.info("Model controller for model {} successfully deleted from index, result: {}", modelId, deleteResponse.getResult());
+                mlModelManager.updateModel(modelId, Map.of(MLModel.IS_MODEL_CONTROLLER_ENABLED_FIELD, false));
```

Comment:
What if this step failed ? Should we add action listener for this method ?

Reply:
You can check the message inside, both the error and response will go into the log. I think it's fine because when doing model deploying I handled this edge case. Once it failed to be updated to false it can never set back to false actively, unless user manually create and delete the controller again.